### PR TITLE
added TableRange IR node

### PR DESCRIPTION
**Pull request by cseed**

### CODE REVIEW

**Comment by cseed** [comment]
... and MatrixRange IR

File: src/main/scala/is/hail/expr/Relational.scala [hunk 1]
```diff
@@ -931,6 +1008,49 @@ case class TableImport(paths: Array[String], typ: TableType, readerOpts: TableRe
   }
 }
 
+case class TableRange(n: Int, name: String = "index", nPartitions: Int) extends TableIR {
```
Comment (jigold) [inline src/main/scala/is/hail/expr/Relational.scala]: Can you have optional parameters before non-optional in Scala?

File: src/main/scala/is/hail/expr/Relational.scala [hunk 2]
```diff
@@ -931,6 +1008,49 @@ case class TableImport(paths: Array[String], typ: TableType, readerOpts: TableRe
...
+  val children: IndexedSeq[BaseIR] = Array.empty[BaseIR]
+
+  def copy(newChildren: IndexedSeq[BaseIR]): TableRange = this
```
Comment (jigold) [inline src/main/scala/is/hail/expr/Relational.scala]: Does this need an assert to be consistent with MatrixRange?

File: src/main/scala/is/hail/expr/Relational.scala [hunk 3]
```diff
@@ -344,6 +344,83 @@ case class MatrixRead(
   override def toString: String = s"MatrixRead($path, dropSamples = $dropCols, dropVariants = $dropRows)"
 }
 
+case class MatrixRange(nRows: Int, nCols: Int, nPartitions: Int) extends MatrixIR {
```
Comment (jigold) [inline src/main/scala/is/hail/expr/Relational.scala]: should we let users specify the row and column names like we do with table?

File: src/main/scala/is/hail/expr/Relational.scala [hunk 4]
```diff
@@ -344,6 +344,83 @@ case class MatrixRead(
...
+    val localPartCounts = partCounts
+    val partStarts = partCounts.scanLeft(0)(_ + _)
+
+    val rvd =
+      OrderedRVD(typ.orvdType,
+        new OrderedRVDPartitioner(typ.rowPartitionKey.toArray,
+          typ.rowKeyStruct,
+          Array.tabulate(nPartitions) { i =>
+            val start = partStarts(i)
+            Interval(start, partCounts(i), includesStart = true, includesEnd = (i == nPartitions - 1))
```
Comment (jigold) [inline src/main/scala/is/hail/expr/Relational.scala]: shouldn't this be `Interval(start, start + partCounts(i), ...)`?

File: src/main/scala/is/hail/expr/Relational.scala [hunk 5]
```diff
@@ -344,6 +344,83 @@ case class MatrixRead(
...
+        hc.sc.parallelize(Range(0, nPartitions), nPartitions)
+          .mapPartitionsWithIndex { case (i, _) =>
+            val region = Region()
+            val rvb = new RegionValueBuilder(region)
+            val rv = RegionValue(region)
+
+            val start = partStarts(i)
+            Iterator.range(start, start + localPartCounts(i))
+              .map { j =>
+                rvb.start(localRVType)
```
Comment (jigold) [inline src/main/scala/is/hail/expr/Relational.scala]: You had `region.clear()` in TableRange. Is this needed?

File: python/hail/tests/test_api.py [hunk 6]
```diff
@@ -556,6 +556,13 @@ def test_str_annotation_regression(self):
         t = t.annotate(ref = t.alleles[0])
         t._force_count()
 
+    def test_range_table(self):
```
Comment (jigold) [inline python/hail/tests/test_api.py]: Do you need the analogous test for MatrixRange here?

File: src/main/scala/is/hail/expr/Relational.scala (hunk 1 above)
Comment (cseed) [reply src/main/scala/is/hail/expr/Relational.scala]: Yes!

```
scala> def f(i: Int, j: Int = 5, k: Int): Int = i + j + k
f: (i: Int, j: Int, k: Int)Int

scala> f(1, 2, 3)
res0: Int = 6

scala> f(1, k=5)
res3: Int = 11
```

However, I admit it is a bit strange and it isn't used, so I'll remove it.

File: src/main/scala/is/hail/expr/Relational.scala (hunk 3 above)
Comment (cseed) [reply src/main/scala/is/hail/expr/Relational.scala]: Table wasn't actually  exposed in Python.  I hardcoded the names everywhere.

**Pull request merged**